web/packages/teleterm/src/mainProcess/agentDownloader/agentDownloader.ts

```diff
@@ -0,0 +1,146 @@
+const logger = new Logger('agentDownloader');
+
+interface AgentBinaryParams {
+  version: string;
+  platform: string;
+  arch: string;
+}
+
+/**
+ * Downloads and unpacks the agent binary, if it has not already been downloaded.
+ * If `RuntimeSettings` does not contain the correct version (so it equals 1.0.0-dev),
+ * the version from AGENT_VERSION env var is taken or the latest available version of the agent is downloaded.
+ */
+export async function downloadAgent(
+  fileDownloader: IFileDownloader,
+  settings: RuntimeSettings
+): Promise<void> {
+  const version = await calculateAgentVersion(settings.appVersion);
+
+  if (await isAgentAlreadyDownloaded(settings.agentBinaryPath, version)) {
+    logger.info(`Agent v.${version} is already downloaded. Skipping.`);
+    return;
+  }
+
+  const binaryName = createAgentBinaryName({
+    arch: settings.arch,
+    platform: settings.platform,
+    version,
+  });
+  const url = `${TELEPORT_CDN_ADDRESS}/${binaryName}`;
+  await fileDownloader.run(url, settings.tempDataDir);
+
+  await unpack(join(settings.tempDataDir, binaryName), settings.sessionDataDir);
+
+  logger.info(`Downloaded agent v.${version}.`);
+}
+
+async function calculateAgentVersion(appVersion: string): Promise<string> {
+  if (appVersion !== '1.0.0-dev') {
+    return appVersion;
+  }
+  if (process.env.AGENT_VERSION) {
+    return process.env.AGENT_VERSION;
+  }
+  return await fetchLatestTeleportRelease();
+}
+
+/**
+ * Takes the first page of teleport releases (30 items) and looks for the highest version.
+ * We don't have a way to simply take the latest tag.
+ */
+async function fetchLatestTeleportRelease(): Promise<string> {
+  const teleportReleases = await fetch(
+    'https://rlz.teleport.sh/teleport?page=0'
+  );
+  const versions = (
+    (await teleportReleases.json()) as {
+      version: string;
+    }[]
+  ).map(r => r.version);
+
+  // get the last element
+  const latest = versions.sort(compareSemVers)?.at(-1);
+  if (latest) {
+    return latest;
+  }
+  throw new Error('Failed to read the latest teleport release.');
+}
+
+/**
+ * Generates following binary names:
+ * teleport-v<version>-linux-arm64-bin.tar.gz
+ * teleport-v<version>-linux-amd64-bin.tar.gz
+ * teleport-v<version>-darwin-arm64-bin.tar.gz
+ * teleport-v<version>-darwin-amd64-bin.tar.gz
+ */
+function createAgentBinaryName(params: AgentBinaryParams): string {
+  const arch = params.arch === 'x64' ? 'amd64' : params.arch;
+  return `teleport-v${params.version}-${params.platform}-${arch}-bin.tar.gz`;
+}
+
+async function isAgentAlreadyDownloaded(
+  agentBinaryPath: string,
+  neededVersion: string
+): Promise<boolean> {
+  const asyncExec = promisify(exec);
+  try {
+    const agentVersion = await asyncExec(
+      `${agentBinaryPath.replace(/ /g, '\\ ')} version --raw`,
+      {
+        timeout: 10_000, // 10 seconds
+      }
+    );
+    return agentVersion.stdout.trim() === neededVersion;
+  } catch (e) {
+    logger.error(e);
+    return false;
```

Comment:
Why do we ignore the error here? I understand that this can help if the file is corrupted but we might as well end up with a bug and unknowingly download a fresh agent on each start.

Reply:
When you run CMC for the first time, the `exec` would fail, because of the missing file. I wanted to inspect if it was a ENOENT error but the returned error was just a string. so I ignored everything 😬

Fortunately, after switching to `execFile` we can check if the error is ENOENT, and re-throw the error in case of the other type of error.

Reply:
Ah, good thing we can now handle ENOENT specifically. Though if we were stuck with `exec`, I'd say we can check for the existence of the file first. ;)